### ISSUE TITLE
[Bug]: Assertion failure `mergedScale.size() == 0` during `runnnlayertests` (Arch Linux, Clang 22.1.3)

### ISSUE THREAD

**Issue by HarangPanul** [lightvector/KataGo#1192]
### Description

I am encountering a fatal assertion failure when running the neural network layer tests on **Arch Linux**. This issue occurs even after stashing all local changes and ensuring a clean build from the master branch.

### Environment

* **OS:** Arch Linux
* **GPU:** NVIDIA GeForce GTX 1650
* **Compiler:** **Clang 22.1.3** (and GCC 16.1.1)
* **Backend:** Reproducible on both **CUDA** and **OpenCL**.
* **Build Type:** Tested in both `Release` and `Debug` modes.

### Steps to Reproduce

1. Build KataGo with the aforementioned environment.
2. Run the command: `./katago runnnlayertests`
3. The program fails during the internal/dummy network test phase.

### Error Logs & Debugging

**Error Message:**

```text
FATAL ERROR: Failed test assert: desc->mergedScale.size() == numChannels
file: .../neuralnet/cudabackend.cpp (or openclbackend.cpp)
line: 540 (or 905)

```

**Manual Inspection:**
I added print statements before the assertion to verify the values:

* `desc->mergedScale.size()`: **0**
* `numChannels`: **2**

It seems that the scaling parameters (mergedScale) are not being initialized or populated into the vector when the internal dummy network is constructed.

### Additional Context

* **No external model:** This happens using the default internal test configuration (no model file provided).
* **Similar issues:** I noticed a similar symptom (0-value outputs followed by a crash) in a recent report regarding Apple Silicon/Metal. While the platforms differ, the core issue of "unitialized/empty NN descriptor data" might be related to how the latest compilers (like Clang 22) handle struct memory layouts or constructor execution.
* The issue persists after `git stash`, confirming it is not caused by my local modifications.

**Comment by HarangPanul** [comment]
I have narrowed down the cause of the mergedScale.size() == 0 assertion failure. Here is a summary of my investigation so far:

    Environment: Arch Linux, CUDA 13.2, Clang 22 / GCC.

    Compilation: Due to the removal of older architecture support in CUDA 13.2, I had to manually modify cpp/CMakeLists.txt to remove architectures below compute_75 (Turing) to get it to compile.

    Test Results:

        v1.16.0: Tests PASSED successfully.

        v1.16.4: Tests FAILED with the mergedScale.size() == 0 error.

    Current Progress: I have confirmed that the regression introduced between v1.16.0 and v1.16.4 is causing the issue in high-version CUDA/Compiler environments. I am currently performing a git bisect between these two versions to identify the exact commit.

I will update this thread as soon as I find the first bad commit.

**Comment by HarangPanul** [comment]
**Subject: Identified first bad commit and root cause analysis**

I believe I have identified the root cause of the `mergedScale.size() == 0` issue. After performing a `git bisect` between **v1.16.0** (Passed) and **v1.16.4** (Failed), I found that the issue was introduced in:

**First Bad Commit:** `80fde3ebc69f2d7b28f7f72edf352cf994595ebf` ("Computed merged bn params in the desc")

**My analysis of the issue:**

In my opinion, the problem stems from a missing initialization path for the merged parameters.

1. **Previous Behavior (Safe):** Before this commit, each backend (CUDA, OpenCL, etc.) computed `mergedScale` locally during its own initialization. This ensured that the values were always calculated regardless of how the layer description was created.
2. **Current Behavior (Potential Leak):** This commit moved the merging logic into the `BatchNormLayerDesc` constructor in `desc.cpp`. However, the calculation is currently nested strictly inside the `istream` constructor.
3. **The Missing Link:** In certain scenarios—such as running `runnnlayertests` or when layers are initialized through default constructors or manual member assignments—the `istream` constructor is bypassed.
4. **Result:** Consequently, `mergedScale` and `mergedBias` remain as empty vectors (size 0). When the backend receives these empty vectors, the `assert(desc->mergedScale.size() == numChannels)` triggers a crash.

I have verified that checking out the commit immediately prior to `80fde3eb` resolves the crash, as the backends revert to local computation. To fix this, the merging logic likely needs to be ensured across all initialization paths in `desc.cpp`.

**Environment:**

* **OS:** Arch Linux
* **Compiler:** Clang 22 / GCC 16
* **CUDA:** 13.2.78 (GTX 1650)

**Comment by lightvector** [comment]
Thanks! I attempted a fix, I think this is now addressed in https://github.com/lightvector/KataGo/commit/c4a594772dbd26ec2f16af265916c7a5ff132eb5

**Comment by HarangPanul** [comment]
Thank you. I've confirmed that both OpenCL and CUDA pass the tests after the pull.
